tensorflow/core/kernels/segment_reduction_ops.cc

```diff
@@ -183,6 +186,54 @@ class SegmentReductionOp : public OpKernel {
   }
 };
 
+#ifdef GOOGLE_CUDA
+//  SegmentSumGPUOp is a segment sum operator implemented for GPU only.
+template <class T, class Index>
+class SegmentSumGPUOp : public OpKernel {
```

Comment:
Hi, the current implementation of SegmentReductionOps is actually rather specialized for non-GPU devices. I'm not sure if the usual different-functor-for-different-device approach is the best way to go here. I actually started this implementation by porting existing code to be reused on GPU but later gave up because it's very inefficient for GPU device; in addition, the original implementation takes data input as non-primitive data types (i.e., TensorMap) which means if I were to use GPU specialized functors, I may have to introduce some extra data types into the .cu file and risk contaminating the .cu file with unnecessary header files; again, with these considerations, I'm not sure overloading SegmentReductionOp is the best approach. Nevertheless, I'm more than happy to consult with your thoughts on this.